container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java

```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
+                                         int index) {
+                super(hit, fieldIterator, index);
+                advanceNext();
+            }
+
+            @Override
+            protected String toValue(Map.Entry<String, Inspector> field) { return field.getKey(); }
+
+        }
+    }
+
+    private static abstract class SummaryIterator<VALUE> implements Iterator<VALUE> {
+
+        private final FastHit hit;
+
+        /** -1 means that the current iterator is the map iterator of the parent hit, not any summary data iterator */
+        private int currentSummaryDataIndex = -1;
+        private Iterator<VALUE> currentIterator;
+        private VALUE previousReturned = null;
+
+        public SummaryIterator(FastHit hit, Iterator<VALUE> mapFieldsIterator) {
+            this.hit = hit;
+            this.currentIterator = mapFieldsIterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (currentIterator.hasNext()) return true;
+            return nextIterator();
+        }
+
+        @Override
+        public VALUE next() {
+            if (currentIterator.hasNext() || nextIterator()) return previousReturned = currentIterator.next();
+            throw new NoSuchElementException();
+        }
+
+        @Override
+        public void remove() {
+            if (previousReturned == null)
+                throw new IllegalStateException();
+            if ( ! ( currentIterator instanceof SummaryData.SummaryDataIterator))
+                currentIterator.remove(); // remove from the map
+            if (hit.removedFields == null)
+                hit.removedFields = new HashSet<>();
+            hit.removedFields.add(nameOf(previousReturned));
+            previousReturned = null;
+        }
+
+        protected abstract String nameOf(VALUE value);
+        protected abstract Iterator<VALUE> iteratorFor(SummaryData summary);
+
+        /** Advanced to the next non-empty iterator, if any */
+        private boolean nextIterator() {
+            while (++currentSummaryDataIndex < hit.summaries.size()) {
+                currentIterator = iteratorFor(hit.summaries.get(currentSummaryDataIndex));
+                if (currentIterator.hasNext())
+                    return true;
+            }
+            return false;
+        }
+
+    }
+
+    private static class FieldIterator extends SummaryIterator<Map.Entry<String, Object>> {
+
+        public FieldIterator(FastHit hit, Iterator<Map.Entry<String, Object>> mapFieldsIterator) {
+            super(hit, mapFieldsIterator);
+        }
+
+        @Override
+        protected String nameOf(Map.Entry<String, Object> value) {
+            return value.getKey();
+        }
+
+        @Override
+        protected Iterator<Map.Entry<String, Object>> iteratorFor(SummaryData summary) {
+            return summary.fieldIterator();
+        }
+
+    }
+
+
+    private static class FieldNameIterator extends SummaryIterator<String> {
```

Comment:
Iterates over all field names in the hit (including all attached SummaryData) using sub-iterators